### PR TITLE
8332082: Shenandoah: Use consistent tests to determine when pre-write barrier is active

### PR DESCRIPTION
**Pull request by earthling-amzn**
This is consistent with c1 and other platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332082](https://bugs.openjdk.org/browse/JDK-8332082): Shenandoah: Use consistent tests to determine when pre-write barrier is active (**Task** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role) ⚠️ Review applies to [2769c97c](https://git.openjdk.org/jdk/pull/19180/files/2769c97cbf5313c5c0f1336060ec39cb66584e3c)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19180/head:pull/19180` \
`$ git checkout pull/19180`

Update a local copy of the PR: \
`$ git checkout pull/19180` \
`$ git pull https://git.openjdk.org/jdk.git pull/19180/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19180`

View PR using the GUI difftool: \
`$ git pr show -t 19180`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19180.diff">https://git.openjdk.org/jdk/pull/19180.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19180#issuecomment-2104884413)

### CODE REVIEW

**Review by kdnilsen — APPROVED** [review]

**Review by shipilev — COMMENTED** [review]
I agree there is an inconsistency, but I also see it is deeper than just these two.
https://github.com/openjdk/jdk/blob/1484153c1a092cefc20270b35aa1e508280843a4/src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp#L696-L698

Before we go hard on either gc-state or SATB "active" flag, let's decide which way we go?

The underlying issue, IIRC, was that hardly any other GC implementation has a GC state flag, so they are forced to use the SATB "active" flags. But I am thinking that testing the gc-state flag is better on these paths, since the gc-state flag is guaranteed to be uncontended and fast-accessible.

**Comment by earthling-amzn** [comment]
Both the gc-state and satb-active flags are synchronized to thread local fields on a safepoint. I expect the cost for the barrier to access either of them is similar. However, there is some additional c2 code that looks like it recognizes the barrier IR based on the presence of nodes to load the gc state: (for example, `ShenandoahBarrierSetC2::is_shenandoah_marking_if`).

**Comment by shipilev** [comment]
Hold on. Here is the C2 SATB barrier check, which checks gc-state for `MARKING`:
https://github.com/openjdk/jdk/blob/2f10a316ff0c5a4c124b94f6fabb38fb119d2c82/src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp#L241-L243

What this PR changes is `ShenandoahBarrierSetAssembler::satb_write_barrier_impl`, which is the SATB barrier for generic assembly code. Yes, some of that may be reached from C2.

Looking around, I see that C1 AArch64, RISC-V, x86 `ShenandoahBarrierSetAssembler::generate_c1_pre_barrier_runtime_stub`-s use `gc_state == MARKING` too:
https://github.com/openjdk/jdk/blob/2f10a316ff0c5a4c124b94f6fabb38fb119d2c82/src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp#L695-L698
https://github.com/openjdk/jdk/blob/2f10a316ff0c5a4c124b94f6fabb38fb119d2c82/src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp#L642-L646
https://github.com/openjdk/jdk/blob/2f10a316ff0c5a4c124b94f6fabb38fb119d2c82/src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp#L226-L228

It looks to me the current status is:

Assembler:
 AArch64, RISC-V: SATBMarkQueue.isActive
 x86, PPC: gc_state == MARKING

C1 (IR):
 All platforms: SATBMarkQueue.isActive

C1 (assembler stub):
 All platforms: gc_state == MARKING

C2 (IR):
 All platforms: gc-state == MARKING

So, are we better off going the other way around, towards gc-state == MARKING? This would "only" need to rewrite AArch64, RISC-V parts in assembler, and shared C1 barrier part.

**Comment by earthling-amzn** [comment]
`ShenandoahBarrierSetC2::verify_gc_barriers` is also looking for IR node pattern for `SATBMarkQueue.isActive`, so it'll need rework too.

File: src/hotspot/share/gc/shenandoah/c1/shenandoahBarrierSetC1.cpp [hunk 1]
```diff
@@ -58,26 +58,32 @@ ShenandoahBarrierSetC1::ShenandoahBarrierSetC1() :
...
+          new LIR_Address(thrd,
+                          in_bytes(ShenandoahThreadLocalData::gc_state_offset()),
+                          T_BYTE);
+  // Read the gc_state flag.
   LIR_Opr flag_val = gen->new_register(T_INT);
-  __ load(mark_active_flag_addr, flag_val);
+  __ load(gc_state_addr, flag_val);
+
+  // Create a mask to test if the marking bit is set.
+  // TODO: can we directly test if bit is set?
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/c1/shenandoahBarrierSetC1.cpp:75]: No, we cannot: C1 LIR does not have a corresponding operation.

**Review by shipilev — COMMENTED** [review]
Dang. So we are regressing C1 performance a little here, by changing the single load-cmp to load-and-cmp. That is unfortunate, but might be acceptable?

**Comment by earthling-amzn** [comment]
We usually benchmark with tiered compilation disabled. I will run some tests with tiered compilation configured to stop at level 1.

**Comment by earthling-amzn** [comment]
Total time here is defined as the sum of 10 measurement runs for each of the 20 working benchmarks in the dacapo suite. The total time was taken for `jdk:master` and this PR with `-XX:TieredStopAtLevel=1` to keep all benchmarks running in C1.
```
jdk:master:
  Category |  Count |         Total |      GeoMean |      Average |     Trim 0.1 |       StdDev |      Minimum |      Maximum
    result |    200 |   2110857.000 |     7656.397 |    10554.285 |     9063.875 |     8894.464 |     1159.000 |    36836.000

This PR:
  Category |  Count |         Total |      GeoMean |      Average |     Trim 0.1 |       StdDev |      Minimum |      Maximum
    result |    200 |   2139852.000 |     7711.781 |    10699.260 |     9129.150 |     9166.330 |     1167.000 |    37336.000
```
This is a 1.4% increase in total time when running with just C1. The increase at trimmed mean 10% is just 0.7%. I'm okay with this.

**Review by shipilev — APPROVED** [review]
Agreed, this is acceptable. We could resolve it in future by expanding gc-state to several bytes.

**Pull request closed without merging**